internal/inspect/policy.go

```diff
@@ -0,0 +1,315 @@
+	responsev1 "github.com/cerbos/cerbos/api/genpb/cerbos/response/v1"
+	runtimev1 "github.com/cerbos/cerbos/api/genpb/cerbos/runtime/v1"
+	"github.com/cerbos/cerbos/internal/compile"
+	"github.com/cerbos/cerbos/internal/namer"
+	"github.com/cerbos/cerbos/internal/policy"
+)
+
+func Policies() *Policy {
+	return &Policy{
+		toResolve: make(map[string]map[string]bool),
+		imports:   make(map[string][]string),
+		results:   make(map[string]*responsev1.InspectPoliciesResponse_Result),
+	}
+}
+
+type Policy struct {
+	toResolve map[string]map[string]bool
+	imports   map[string][]string
+	results   map[string]*responsev1.InspectPoliciesResponse_Result
+}
+
+// Inspect inspects the given policy and caches the inspection related information internally.
+func (pol *Policy) Inspect(p *policyv1.Policy) error {
+	if p == nil {
+		return fmt.Errorf("policy is nil")
+	}
+
+	policyID := namer.PolicyKey(p)
+	variables := policy.ListVariables(p)
+	if _, ok := p.PolicyType.(*policyv1.Policy_ExportVariables); ok {
+		sort.Slice(variables, func(i, j int) bool {
+			return variables[i].Name < variables[j].Name
+		})
+
+		if len(variables) > 0 {
+			pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
+				Variables: variables,
+			}
+		}
+
+		return nil
+	}
+
+	referencedVariables, err := pol.inspectConditions(p)
+	if err != nil {
+		return fmt.Errorf("failed to inspect conditions of the policy %s: %w", policyID, err)
+	}
+	pol.imports[policyID] = pol.inspectImports(p)
+
+	localVariables := make(map[string]struct{})
+	for _, variable := range variables {
+		localVariables[variable.Name] = struct{}{}
+		if _, ok := referencedVariables[variable.Name]; ok {
+			variable.Used = true
+		}
+	}
+
+	toResolve := false
+	for name := range referencedVariables {
+		if _, ok := localVariables[name]; !ok {
+			if pol.toResolve[policyID] == nil {
+				pol.toResolve[policyID] = make(map[string]bool)
+			}
+
+			toResolve = true
+			pol.toResolve[policyID][name] = false
+		}
+	}
+
+	// sort variables if there is nothing to resolve since we are not going to modify variables in the future.
+	if !toResolve {
+		sort.Slice(variables, func(i, j int) bool {
+			return variables[i].Name < variables[j].Name
+		})
+	}
+
+	a := policy.ListActions(p)
+	sort.Strings(a)
+	if len(a) > 0 || len(variables) > 0 {
+		pol.results[policyID] = &responsev1.InspectPoliciesResponse_Result{
+			Actions:   a,
+			Variables: variables,
+		}
+	}
+
+	return nil
+}
+
+// Results returns the final inspection results.
+func (pol *Policy) Results() (map[string]*responsev1.InspectPoliciesResponse_Result, error) {
+	for policyID, variables := range pol.toResolve {
+		importedPolicies, ok := pol.imports[policyID]
+		if ok {
+			for _, importedPolicyID := range importedPolicies {
+				importedResult, ok := pol.results[importedPolicyID]
+				if !ok {
+					return nil, fmt.Errorf("failed to find imported policy %s in the inspected policies", importedPolicyID)
```

Comment:
We should try to load this policy from the store. It's quite possible that the list of policies the user asked to inspect doesn't contain all the required `ExportVariables` policies. It's not an error.

- Go through the list of policies with unresolved variables
- Collect their imports into a set
- Load that set from the store and try to resolve the unresolved variables again
- Anything remaining unresolved is an undefined variable

Reply:
I do something like this with the dbStorage. The steps:
- Call `ListPolicyIDs` with filters
  - For each policy, call `Inspect` method which also caches the imported policy names into `pol.imports`
- After the list of policies are exhausted, call the [MissingImports](https://github.com/cerbos/cerbos/pull/2141/files#diff-9fe34d6e3fe21b240b8b0e0e84996b570289e47dd03e655d5e9f44dcd08d8fd1R100-R117) method to get policies left outside of filtered policies
- [Call LoadPolicy with the list of missing policies and call Inspect on them](https://github.com/cerbos/cerbos/pull/2141/files#diff-1823d40180fc37bc05da9d1175fee7d28d4ccad65bde1542981f12e4cfe30782R677)
- Call and return `Results`

Reply:
I saw that. There are a couple of issues with that.
- It implicitly adds the missing imported variables to your result. The user didn't ask for those so it's surprising behaviour.
- The logic is now in multiple places. You have to remember to call `MissingImports` and inspect those too before calling `Result`. 

The missing imports are a special case. They shouldn't be added the final result as "policies". We only need them for resolving variables we found while inspecting the original set of policies. I'd add a parameter to the `Result` method. It'd be a function that takes the list of missing imports and returns the variables defined in those imports.